canary/S3ObjectTransport.cpp

```diff
@@ -0,0 +1,916 @@
+#include <iostream>
+
+#if defined(_WIN32)
+#    undef min
+#endif
+
+using namespace Aws::Crt;
+
+const uint32_t S3ObjectTransport::MaxUploadMultipartStreams = 800;
+const uint32_t S3ObjectTransport::MaxDownloadMultipartStreams = 800;
+
+namespace
+{
+    const uint32_t TransfersPerAddress = 10;
+    const int32_t S3GetObjectResponseStatus_PartialContent = 206;
+} // namespace
+
+S3ObjectTransport::S3ObjectTransport(CanaryApp &canaryApp, const Aws::Crt::String &bucket)
+    : m_canaryApp(canaryApp), m_bucketName(bucket), m_connManagersUseCount(0), m_activeRequestsCount(0),
+      m_uploadProcessor(canaryApp, canaryApp.GetEventLoopGroup(), MaxUploadMultipartStreams),
+      m_downloadProcessor(canaryApp, canaryApp.GetEventLoopGroup(), MaxDownloadMultipartStreams)
+{
+    m_endpoint = m_bucketName + ".s3." + m_canaryApp.GetOptions().region.c_str() + ".amazonaws.com";
+
+    m_hostHeader.name = ByteCursorFromCString("host");
+    m_hostHeader.value = ByteCursorFromCString(m_endpoint.c_str());
+
+    m_contentTypeHeader.name = ByteCursorFromCString("content-type");
+    m_contentTypeHeader.value = ByteCursorFromCString("text/plain");
+}
+
+void S3ObjectTransport::EmitS3AddressCountMetric(size_t addressCount)
+{
+    AWS_LOGF_INFO(AWS_LS_CRT_CPP_CANARY, "Emitting S3 Address Count Metric: %" PRIu64, (uint64_t)addressCount);
+
+    Metric s3AddressCountMetric(MetricName::S3AddressCount, MetricUnit::Count, 0ULL, (double)addressCount);
+    m_canaryApp.GetMetricsPublisher()->AddDataPoint(s3AddressCountMetric);
+}
+
+void S3ObjectTransport::WarmDNSCache(uint32_t numTransfers)
+{
+    // Each transfer is in a group the size of TransfersPerAddress,
+    size_t desiredNumberOfAddresses = numTransfers / TransfersPerAddress;
+
+    if ((numTransfers % TransfersPerAddress) > 0)
+    {
+        ++desiredNumberOfAddresses;
+    }
+
+    AWS_LOGF_INFO(
+        AWS_LS_CRT_CPP_CANARY,
+        "Warming DNS cache: getting %" PRIu64 " addresses for endpoint %s",
+        (uint64_t)desiredNumberOfAddresses,
+        m_endpoint.c_str());
+
+    // Ask the host resolver to start resolving.
+    m_canaryApp.GetDefaultHostResolver().ResolveHost(
+        m_endpoint, [](Io::HostResolver &, const Vector<Io::HostAddress> &, int) {});
+
+    // Wait until the resolved address count is what we need it to be.
+    {
+        size_t numAddresses = m_canaryApp.GetDefaultHostResolver().GetHostAddressCount(
+            m_endpoint, AWS_GET_HOST_ADDRESS_COUNT_RECORD_TYPE_A);
+
+        EmitS3AddressCountMetric(numAddresses);
+
+        while (numAddresses < desiredNumberOfAddresses)
+        {
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+
+            numAddresses = m_canaryApp.GetDefaultHostResolver().GetHostAddressCount(
+                m_endpoint, AWS_GET_HOST_ADDRESS_COUNT_RECORD_TYPE_A);
+
+            EmitS3AddressCountMetric(numAddresses);
+        }
+    }
+
+    // Clear our local cache in preparation for getting address from the host resolver.
```

Comment:
Probably worth noting that the reason you throw these away is to let the host resolver manage TTLs

Reply:
It's actually because we're building up our own cache at this point to divide them among multiple connection managers.  At some point we should be able to do away with some of this setup and hopefully just have the host resolver handle a lot of this.